crates/bevy_ecs/src/query/access_iter.rs

```diff
@@ -0,0 +1,521 @@
+            | (Component(FilteredWriteAll), _)
+            | (_, Component(FilteredWriteAll)) => {
+                AccessCompatible::Compatible
+            }
+
+            // single component checks
+            (Component(Read(id)), Component(Write(id_other)))
+            | (Component(Write(id)), Component(Read(id_other)))
+            | (Component(Write(id)), Component(Write(id_other)))
+            | (
+                Resource(ResourceAccessLevel::Read(id)),
+                Resource(ResourceAccessLevel::Write(id_other)),
+            )
+            | (
+                Resource(ResourceAccessLevel::Write(id)),
+                Resource(ResourceAccessLevel::Read(id_other)),
+            )
+            | (
+                Resource(ResourceAccessLevel::Write(id)),
+                Resource(ResourceAccessLevel::Write(id_other)),
+            ) => (id != id_other).into(),
+
+            // Except* access in first parameter
+            (
+                Component(ReadAllExcept {
+                    component_id: id,
+                    index,
+                }),
+                Component(Write(id_other)),
+            )
+            | (
+                Component(WriteAllExcept {
+                    component_id: id,
+                    index,
+                }),
+                Component(Read(id_other)),
+            )
+            | (
+                Component(WriteAllExcept {
+                    component_id: id,
+                    index,
+                }),
+                Component(Write(id_other)),
+            )
+            // Except* access in second parameter
+            | (
+                Component(Write(id)),
+                Component(ReadAllExcept {
+                    component_id: id_other,
+                    index,
+                }),
+            )
+            | (
+                Component(Read(id)),
+                Component(WriteAllExcept {
+                    component_id: id_other,
+                    index,
+                }),
+            )
+            | (
+                Component(Write(id)),
+                Component(WriteAllExcept {
+                    component_id: id_other,
+                    index,
+                }),
+            ) => {
+                if id == id_other {
+                    AccessCompatible::CompatibleExcept(index)
+                } else {
+                    AccessCompatible::ConflictsExcept(index)
+                }
+            }
+
+            // WriteAll will always conflict if they have different indexes
+            (
+                Component(WriteAllExcept { index, .. }),
+                Component(WriteAllExcept {
+                    index: index_other, ..
+                }),
+            ) => (index == index_other).into(),
+        }
+    }
+}
+
+/// Error returned from [`has_conflicts`].
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum QueryAccessError {
+    /// Component was not registered on world
+    ComponentNotRegistered,
```

Comment:
Note that since adding a component to an entity registers it, `ComponentNotRegistered` is just a special case of `EntityDoesNotMatch`.  I don't think a caller can ever really do anything differently in those cases, so I'd be inclined to remove the `ComponentNotRegistered` variant and just return `EntityDoesNotMatch` in that case.